src/libstd/sys/windows/path.rs

```diff
@@ -82,11 +82,33 @@ pub fn parse_prefix(path: &OsStr) -> Option<Prefix<'_>> {
         return None;
     }
 
-    fn parse_two_comps(mut path: &[u8], f: fn(u8) -> bool) -> Option<(&[u8], &[u8])> {
-        let first = &path[..path.iter().position(|x| f(*x))?];
-        path = &path[(first.len() + 1)..];
-        let idx = path.iter().position(|x| f(*x));
-        let second = &path[..idx.unwrap_or(path.len())];
+    /// Split path component that matches predicate `f`.
+    ///
+    /// Returns [`None`] if there is no separator in path.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::::{split_path_component, is_verbatim_sep};
+    /// assert_eq!(
+    ///     split_path_component(br"server\share", is_verbatim_sep),
+    ///     Some((&b"server"[..], &b"share"[..])),
+    /// );
+    /// assert_eq!(
+    ///     split_path_component(br"server\", is_verbatim_sep),
+    ///     Some((&b"server"[..], &b""[..]))
+    /// );
+    /// ```
+    pub fn split_path_component<'a>(
+        path: &'a [u8],
+        f: fn(u8) -> bool,
+    ) -> Option<(&'a [u8], &'a [u8])> {
+        let idx = path.iter().position(|&x| f(x))?;
+        // Remove the path component separator
+        // SAFETY: `s[idx + 1..] == s[s.len()..]`
```

Comment:
That safety comment is confusing IMO. How about something like this?

```
// SAFETY: `idx + 1` is never larger than `path.len()` and `path[path.len()..]` 
// is a valid index
```

Either way, is the `unsafe` even worth it here? Might this be in a hot loop? I have no idea, but if it's unlikely this function is performance critical, I'd rather avoid the `unsafe` code.

Reply:
I know it is confusing but I couldn't think of any better comments for it.
This is really nice spot.

Re `unsafe`, I was haunt by #73396 and outer unsafe blocks made me feel too powerful to use unsafe here.
Anyway, Nikita assigned to that issue and it would be resolve by next LLVM update, I will drop the unsafe indexing.